poc/tenant-controller/pkg/controllers/tenants/controller.go

```diff
@@ -67,63 +74,174 @@ func (c *Controller) Run(ctx context.Context) error {
+	}
+	sort.Strings(oldNsList)
+	nsList := make([]*tenantsapi.TenantNamespace, len(obj.Spec.Namespaces))
+	for i := range obj.Spec.Namespaces {
+		nsList[i] = &obj.Spec.Namespaces[i]
+	}
+	sort.Slice(nsList, func(i, j int) bool {
+		return strings.Compare(nsList[i].Name, nsList[j].Name) < 0
+	})
+	var i, j int
+	for i < len(oldNsList) && j < len(nsList) {
+		if res := strings.Compare(oldNsList[i], nsList[j].Name); res == 0 {
+			c.syncNamespaceForTenant(obj, nsList[j])
+			i++
+			j++
+		} else if res < 0 {
+			c.deleteNamespaceForTenant(obj, oldNsList[i])
+			i++
+		} else {
+			c.createNamespaceForTenant(obj, nsList[j])
+			j++
+		}
+	}
 
-	namespace := corev1.Namespace{}
+	for ; j < len(nsList); j++ {
+		c.createNamespaceForTenant(obj, nsList[j])
+	}
+	for ; i < len(oldNsList); i++ {
+		c.deleteNamespaceForTenant(obj, oldNsList[i])
+	}
+}
+
+func (c *Controller) deleteTenant(obj *tenantsapi.Tenant) {
+	glog.V(2).Infof("deleteTenant: %#v", obj)
 
-	for n := range obj.Spec.Namespaces {
-		namespace.ObjectMeta.Name = obj.Spec.Namespaces[n].Name
-		//Create namespace
-		c.k8sclient.CoreV1().Namespaces().Create(&namespace)
+	// TODO with OwnerReferences, no extra work is needed in deletion,
+	// remove the following code later.
 
-		glog.Info("Created namespace: %s", obj.Spec.Namespaces[n].Name)
+	c.deleteRBACForTenant(obj)
+	for _, nsReq := range obj.Spec.Namespaces {
+		c.deleteNamespaceForTenant(obj, nsReq.Name)
 	}
-
-	glog.Info("createTenant completed: %#v", obj)
-	// TODO create RBAC inside namespace
+	glog.Infof("deleteTenant completed: %#v", obj)
 }
 
-func (c *Controller) updateTenant(old, obj *tenantsapi.Tenant) {
-	glog.Info("updateTenant: %#v", obj)
-	syncRBACForTenantCRD(obj)
-	// TODO sync namespace
-	// TODO sync RBAC inside namespace
+func (c *Controller) createNamespaceForTenant(tenant *tenantsapi.Tenant, nsReq *tenantsapi.TenantNamespace) error {
+	if err := c.ensureNamespaceExists(tenant, nsReq.Name); err != nil {
+		return err
+	}
+	if err := c.syncNamespaceForTenant(tenant, nsReq); err != nil {
+		return err
+	}
+	glog.Infof("Tenant %q namespace %q created", tenant.Name, nsReq.Name)
+	return nil
 }
 
-func (c *Controller) deleteTenant(obj *tenantsapi.Tenant) {
-	glog.Info("deleteTenant: %#v", obj)
-	deleteRBACForTenantCRD(obj)
+func (c *Controller) deleteNamespaceForTenant(tenant *tenantsapi.Tenant, nsName string) error {
 	// TODO add a full set of sanity checks in future before deleting
+	if err := c.k8sclient.CoreV1().Namespaces().Delete(nsName, nil); err != nil {
+		glog.Errorf("Tenant %q delete namespace %q error: %v", tenant.Name, nsName, err)
+		return err
+	}
+	glog.Infof("Tenant %q namespace %q deleted", tenant.Name, nsName)
+	return nil
+}
+
+func (c *Controller) ensureNamespaceExists(tenant *tenantsapi.Tenant, nsName string) error {
+	// TODO Add later ... sanity checks to ensure namespaces being requested are valid and not already assigned to another tenant
+	if err := newKubeCtl().addObjects(&corev1.Namespace{
```

Comment:
Minor comment: In the same file we now have calls using k8sclient as well as newKubeCtl() ... we should discuss whether to make it consistent and use the same approach everywhere or whether it is okay to use different approaches.  The separate k8sclient approach was taken to natch the example in the sample-controller repo.